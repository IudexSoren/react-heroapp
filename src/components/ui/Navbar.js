import { useContext } from 'react';
import { Link, NavLink, useHistory } from 'react-router-dom'
import { AuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';

const Navbar = () => {

    const { user:{ name }, dispatch } = useContext(AuthContext);

    // Navbar tiene acceso al hook useHistory al estar dentro de ContextProvider
    const history = useHistory();

    const handleLogout = () => {
        const action = {
            type: types.logout
        }
        dispatch(action);
        history.replace('/login');
    }

    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark px-2">
            <Link
                className="navbar-brand"
                to="/"
            >
                Associations
            </Link>
            <div className="navbar-collapse">
                <div className="navbar-nav">
                    <NavLink
                        activeClassName="active"
                        className="nav-item nav-link"
                        exact
                        to="/marvel"
                    >
                        Marvel
                    </NavLink>
                    <NavLink
                        activeClassName="active"
                        className="nav-item nav-link"
                        exact
                        to="/dc"
                    >
                        DC
                    </NavLink>
                    <NavLink
                        activeClassName="active"
                        className="nav-item nav-link"
                        exact
                        to="/search"
                    >
                        SearchðŸ”Ž
                    </NavLink>
                </div>
            </div>
            <div>
                <ul className="navbar-nav ml-auto">
                    <span className="nav-item nav-link text-info">{ name }</span>
                    <button
                        className="nav-item nav-link btn"
                        onClick={ handleLogout }
                    >
                        Logout
                    </button>
                </ul>
            </div>
        </nav>
    )
}

export default Navbar;