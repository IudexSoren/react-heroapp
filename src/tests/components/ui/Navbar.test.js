import { mount } from "enzyme";
import { MemoryRouter, Router } from "react-router-dom";
import { AuthContext } from "../../../auth/AuthContext";
import Navbar from "../../../components/ui/Navbar";
import { types } from "../../../types/types";


describe('Pruebas de Navbar', () => {

  const historyMock = {
    replace: jest.fn(),
    push: jest.fn(),
    location: {},
    listen: jest.fn(),
    createHref: jest.fn()
  }

  const contextValue = {
    dispatch: jest.fn(),
    user: {
      name: 'Aar√≥n',
      logged: true
    }
  }

  const wrapper = mount(
    <AuthContext.Provider value={ contextValue } >
      <MemoryRouter>
        <Router history={ historyMock }>
          <Navbar />
        </Router>
      </MemoryRouter>
    </AuthContext.Provider>
  );

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Debe renderizar correctamente el componente y mostrar el nombre del usuario autenticado', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('span').text()).toBe(contextValue.user.name);
  });

  test('Debe ejecutar el logout y history.replace', () => {
    wrapper.find('button').simulate('click');

    expect(contextValue.dispatch).toHaveBeenCalledWith({
      type: types.logout
    });
    expect(historyMock.replace).toHaveBeenCalledWith('/login');
  });

});
