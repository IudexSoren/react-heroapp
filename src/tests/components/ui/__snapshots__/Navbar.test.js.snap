// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de Navbar Debe renderizar correctamente el componente y mostrar el nombre del usuario autenticado 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark px-2"
>
  <a
    className="navbar-brand"
    onClick={[Function]}
  >
    Associations
  </a>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        aria-current={null}
        className="nav-item nav-link"
        onClick={[Function]}
      >
        SearchðŸ”Ž
      </a>
    </div>
  </div>
  <div>
    <ul
      className="navbar-nav ml-auto"
    >
      <span
        className="nav-item nav-link text-info"
      >
        AarÃ³n
      </span>
      <button
        className="nav-item nav-link btn"
        onClick={[Function]}
      >
        Logout
      </button>
    </ul>
  </div>
</nav>
`;
